src/source/Ice/ConnectionListener.c

```diff
@@ -56,31 +62,35 @@ STATUS freeConnectionListener(PConnectionListener* ppConnectionListener)
     ATOMIC_STORE_BOOL(&pConnectionListener->terminate, TRUE);
 
     if (IS_VALID_MUTEX_VALUE(pConnectionListener->lock)) {
-        // Try to await for the thread to finish up
-        // NOTE: As TID is not atomic we need to wrap the read in locks
-        timeToWait = GETTIME() + CONNECTION_LISTENER_SHUTDOWN_TIMEOUT;
-
-        do {
-            MUTEX_LOCK(pConnectionListener->lock);
-            threadId = pConnectionListener->receiveDataRoutine;
-            MUTEX_UNLOCK(pConnectionListener->lock);
-            if (!IS_VALID_TID_VALUE(threadId)) {
-                threadTerminated = TRUE;
-            }
+        MUTEX_LOCK(pConnectionListener->lock);
+        threadId = pConnectionListener->receiveDataRoutine;
+        MUTEX_UNLOCK(pConnectionListener->lock);
 
-            // Allow the thread to finish and exit
-            if (!threadTerminated) {
-                THREAD_SLEEP(KVS_ICE_SHORT_CHECK_DELAY);
-            }
-        } while (!threadTerminated && GETTIME() < timeToWait);
```

Comment:
Curious. What caused this block to crash with the timeout approach? Also since this is not available in Windows, there is a chance of crash on Windows correct?

Reply:
1. The crash was caused from the connectionListenerReceiveDataRoutine access the ConnectionListener object after the destructor has freed it. The thread would still be running after the timeout occurs.

2. No, the thread_join() is the crux of the fix, since it will have the destructor wait indefinitely for the thread to close. The part windows is missing is the socketpair() kicking poll() out early, so in the worst case windows will have to wait 200ms more.